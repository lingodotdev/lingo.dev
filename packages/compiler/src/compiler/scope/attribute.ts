import * as t from '@babel/types';
import { NodePath } from '@babel/core';
import { IReplexicaScope } from "./types";
import { ReplexicaChunk } from './chunk';
import { generateScopeId } from '../../utils/id';
import { findJsxParentForTheAttribute, getDefaultImportName, getImportName, getJsxAttributeName, getJsxElementName, getStringAttributeValue, injectDefaultImport, injectImport, parseMemberExpressionFromJsxMemberExpression } from '../../utils/ast';
import { ReplexicaScopeData, ReplexicaScopeHint } from '../types';
import { ReplexicaBaseScope } from './base';

// TODO: @/i18n should be a relative import, in case
// @ isn't mapped to the project root

export class ReplexicaAttributeScope extends ReplexicaBaseScope implements IReplexicaScope {
  public static fromNode(path: NodePath<t.Node>): IReplexicaScope[] {
    const result: IReplexicaScope[] = [];
    if (!path.isJSXElement()) { return result; }

    const supportedAttributeNames = ['title', 'alt', 'placeholder'];
    const openingElement = path.get('openingElement');
    const attributesPaths = openingElement.get('attributes');
    for (const attributePath of attributesPaths) {
      if (!attributePath.isJSXAttribute()) { continue; }

      const attributeName = attributePath.get('name');
      if (!attributeName.isJSXIdentifier()) { continue; }

      const attributeNameValue = attributeName.node.name;
      if (!supportedAttributeNames.includes(attributeNameValue)) { continue; }

      result.push(new ReplexicaAttributeScope(attributePath));
    }

    return result;
  }

  public constructor(
    private readonly path: NodePath<t.JSXAttribute>,  
  ) {
    super();

    const valuePath = this.path.get("value") as NodePath<t.Node>;
    if (valuePath.isStringLiteral()) {
      this._chunk = ReplexicaChunk.fromStringLiteral(valuePath);
    } else if (valuePath.isJSXExpressionContainer()) {
      const expressionPath = valuePath.get("expression") as NodePath<t.Node>;
      if (expressionPath.isStringLiteral()) {
        this._chunk = ReplexicaChunk.fromJsxExpressionContainer(valuePath);
      }
    }

    const chunkIds = this._chunk ? [this._chunk.id] : [];
    const result = generateScopeId(chunkIds, 0);
    this._id = result;
  }

  private _chunk: ReplexicaChunk | null = null;
  private _id: string;

  public get id(): string {
    return this._id;
  }

  public injectIntl(fileId: string, isServer: boolean): ReplexicaScopeData {
    const result: ReplexicaScopeData = {};
    if (!this._chunk) { return result; }

    const programNode = this.path.findParent((p) => p.isProgram()) as NodePath<t.Program> | null;
    if (!programNode) { throw new Error(`Couldn't find file node`); }

    const jsxElement = findJsxParentForTheAttribute(this.path);
    if (!jsxElement) {
      throw new Error(`Couldn't find JSX element for attribute ${this.path.get("name").toString()}`);
    }

    const packageName = isServer ? '@replexica/react/next' : '@replexica/react/client';
    const localHelperName = 'I18nProxy';

    let helperName = getImportName(programNode, packageName, localHelperName);
    if (!helperName) {
      helperName = injectImport(programNode, packageName, localHelperName);
    }

    const isProxyAlreadyApplied = t.isJSXIdentifier(jsxElement.node.openingElement.name) && jsxElement.node.openingElement.name.name === helperName;
    if (!isProxyAlreadyApplied) {
      // 1. add $$Component attribute to the element and set it to the original component name / identifier / member expression
      if (t.isJSXIdentifier(jsxElement.node.openingElement.name)) {
        // if it's an identifier, it's a native element or a component
        if (/^[a-z]/.test(jsxElement.node.openingElement.name.name)) {
          // if it's lowercased identifier, it's a native element
          jsxElement.node.openingElement.attributes.push(
            t.jSXAttribute(
              t.jSXIdentifier('$$Component'),
              t.stringLiteral(jsxElement.node.openingElement.name.name),  
            ),
          );
        } else {
          // if it's uppercased identifier, it's a component
          jsxElement.node.openingElement.attributes.push(
            t.jSXAttribute(
              t.jSXIdentifier('$$Component'),
              t.jSXExpressionContainer(t.identifier(jsxElement.node.openingElement.name.name)),
            ),
          );
        }
      } else if (t.isJSXMemberExpression(jsxElement.node.openingElement.name)) {
        const memberExpression = parseMemberExpressionFromJsxMemberExpression(jsxElement.node.openingElement.name);
        jsxElement.node.openingElement.attributes.push(
          t.jSXAttribute(
            t.jSXIdentifier('$$Component'),
            t.jSXExpressionContainer(memberExpression),
          ),
        );
      } else {
        throw new Error(`Unsupported JSXElement name type.`);
      }
      // 2. replace the element name with the name of the proxy component
      jsxElement.node.openingElement.name = t.jSXIdentifier(helperName);
      if (jsxElement.node.closingElement) {
        jsxElement.node.closingElement.name = t.jSXIdentifier(helperName);
      }
    }

    // add the current attribute to the proxy component's $$Attributes attribute
    // like so: <ProxyComponent $$Attributes={{ [attributeName]: { fileId, scopeId, chunkId } }} />
    // create $$Attributes attribute entry only if it doesn't exist yet
    let systemReplexicaAttribute = jsxElement.node.openingElement.attributes.find((attr) => {
      return t.isJSXAttribute(attr) && t.isJSXIdentifier(attr.name) && attr.name.name === '$$Attributes';
    }) as t.JSXAttribute | undefined;
    if (!systemReplexicaAttribute) {
      systemReplexicaAttribute = t.jSXAttribute(
        t.jSXIdentifier('$$Attributes'),
        t.jSXExpressionContainer(t.objectExpression([])),
      );
      jsxElement.node.openingElement.attributes.push(systemReplexicaAttribute);
    }

    // create the current attribute entry
    const attributeName = this.path.get("name").toString();

    const selectorProperty = t.objectProperty(
      t.stringLiteral(attributeName),
      t.objectExpression([
        t.objectProperty(t.stringLiteral('fileId'), t.stringLiteral(fileId)),
        t.objectProperty(t.stringLiteral('scopeId'), t.stringLiteral(this.id)),
        t.objectProperty(t.stringLiteral('chunkId'), t.stringLiteral(this._chunk.id)),
      ]),
    );
    // add the current attribute entry to the $$Attributes attribute
    if (t.isJSXExpressionContainer(systemReplexicaAttribute.value) && t.isObjectExpression(systemReplexicaAttribute.value.expression)) {
      systemReplexicaAttribute.value.expression.properties.push(selectorProperty);
      
      if (isServer) {
        // make sure the following import is available in the file:
        // import i18n from '@/i18n';
        let localeDataLoaderImportName = getDefaultImportName(programNode, '@/i18n');
        if (!localeDataLoaderImportName) {
          localeDataLoaderImportName = injectDefaultImport(programNode, '@/i18n', 'loadI18n');
        }
        
        // add the following props to the injected element:
        // loadLocaleData={localeDataLoaderImportName.loadData}
        jsxElement.node.openingElement.attributes.push(
          t.jsxAttribute(
            t.jsxIdentifier('loadI18n'),
            t.jsxExpressionContainer(
              t.identifier(localeDataLoaderImportName),
            ),
          )
        );
      }

      result[this._chunk.id] = this._chunk.text;
    }

    return result;
  }

  public extractHints(): ReplexicaScopeHint[] {
    const jsxElement = findJsxParentForTheAttribute(this.path);
    if (!jsxElement) {
      throw new Error(`Couldn't find JSX element for attribute ${this.path.get("name").toString()}`);
    }

    const attributeName = getJsxAttributeName(this.path);
    const attributeHint: ReplexicaScopeHint = {
      type: 'attribute',
      name: attributeName,
    };

    const baseHints = this._extractBaseHints(jsxElement);

    const result = [...baseHints, attributeHint];
    return result;
  }
}
