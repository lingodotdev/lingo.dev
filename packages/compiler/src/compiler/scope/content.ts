import * as t from '@babel/types';
import { NodePath } from '@babel/core';
import { IReplexicaScope } from "./types";
import { findImmediateJsxParent, getImportName, hasJsxTextChildren, injectImport } from './../../utils/ast';
import { ReplexicaChunk } from './chunk';
import { generateScopeId } from '../../utils/id';
import { ReplexicaScopeData, ReplexicaScopeHint } from '../types';
import { ReplexicaBaseScope } from './base';

export class ReplexicaContentScope extends ReplexicaBaseScope implements IReplexicaScope {
  public static fromNode(path: NodePath<t.Node>): IReplexicaScope[] {
    if (!path.isJSXElement() && !path.isJSXFragment()) { return []; }
    // to return true, must have non-empty when trimmed JSXText children
    // and either not have a parent JSX element at all, or have a parent JSX element with no text children
    const hasTextContent = hasJsxTextChildren(path);
    if (!hasTextContent) { return []; }

    const jsxElementContainer = findImmediateJsxParent(path);
    if (jsxElementContainer && hasJsxTextChildren(jsxElementContainer)) { return []; }

    const scope = new ReplexicaContentScope(path);
    return [scope];
  }

  private constructor(
    private readonly path: NodePath<t.JSXElement | t.JSXFragment>,
  ) {
    super();
    const _scope = this;

    path.traverse({
      JSXOpeningElement(path) {
        path.skip();
      },
      JSXText(path) {
        const chunk = ReplexicaChunk.fromJsxText(path);
        if (chunk.text.length) {
          _scope._chunks.add(chunk);
        }
      },
      JSXExpressionContainer(path) {
        const chunk = ReplexicaChunk.fromJsxExpressionContainer(path);
        if (chunk.text.length) {
          _scope._chunks.add(chunk);
        }
      },
    });

    const chunkIds = Array.from(this._chunks).map((chunk) => chunk.id);
    this._id = generateScopeId(chunkIds, 0);
  }

  private _chunks: Set<ReplexicaChunk> = new Set();
  private _id: string;

  public get id(): string {
    return this._id;
  }

  public injectIntl(fileId: string, isServer: boolean): ReplexicaScopeData {
    const result: ReplexicaScopeData = {};

    const programNode = this.path.findParent((p) => p.isProgram()) as NodePath<t.Program> | null;
    if (!programNode) { throw new Error(`Couldn't find file node`); }

    const packageName = isServer ? '@replexica/react/server' : '@replexica/react/client';
    const localHelperName = isServer ? 'ReplexicaServerChunk' : 'ReplexicaClientChunk';

    for (const chunk of this._chunks) {
      if (chunk.isPlaceholder) { continue; }

      let helperName = getImportName(programNode, packageName, localHelperName);
      if (!helperName) {
        helperName = injectImport(programNode, packageName, localHelperName);
      }

      const injectedElement = t.jsxOpeningElement(
        t.jsxIdentifier(helperName),
        [
          t.jsxAttribute(t.jsxIdentifier('fileId'), t.stringLiteral(fileId)),
          t.jsxAttribute(t.jsxIdentifier('scopeId'), t.stringLiteral(this.id)),
          t.jsxAttribute(t.jsxIdentifier('chunkId'), t.stringLiteral(chunk.id)),
        ],
        true,
      );

      if (isServer) {
        // make sure the following import is available in the file:
        // import { loadLocaleFromCookie } from '@replexica/react/next';
        let localeLoaderImportName = getImportName(programNode, '@replexica/react/next', 'loadLocaleFromCookie');
        if (!localeLoaderImportName) {
          localeLoaderImportName = injectImport(programNode, '@replexica/react/next', 'loadLocaleFromCookie');
        }
        // add the following props to the injected element:
        // loadLocale={loadLocaleFromCookie}
        injectedElement.attributes.push(
          t.jsxAttribute(
            t.jsxIdentifier('loadLocale'), 
            t.jsxExpressionContainer(t.identifier(localeLoaderImportName)),
          )
        );
        // make sure the following import is available in the file:
        // import { loadLocaleData } from '@replexica/react/next';
        let localeDataLoaderImportName = getImportName(programNode, '@replexica/react/next', 'loadLocaleData');
        if (!localeDataLoaderImportName) {
          localeDataLoaderImportName = injectImport(programNode, '@replexica/react/next', 'loadLocaleData');
        }
        // add the following props to the injected element:
        // loadLocaleData={(locale) => loadLocaleData(i18nImportPrefix, locale)}
        // injectedElement.attributes.push(
        //   t.jsxAttribute(
        //     t.jsxIdentifier('loadLocaleData'), 
        //     t.jsxExpressionContainer(
        //       t.arrowFunctionExpression(
        //         [t.identifier('locale')],
        //         t.callExpression(
        //           t.identifier(localeDataLoaderImportName),
        //           [t.stringLiteral(i18nImportPrefix), t.identifier('locale')],
        //         ),
        //       ),
        //     ),
        //   )
        // );

        // add the following props to the injected element:
        // loadLocaleData={(locale) => import(`@replexica/translations/${locale}.json`)}
        injectedElement.attributes.push(
          t.jsxAttribute(
            t.jsxIdentifier('loadLocaleData'), 
            t.jsxExpressionContainer(
              t.arrowFunctionExpression(
                [t.identifier('locale')],
                t.callExpression(
                  t.identifier('import'),
                  [t.templateLiteral([
                    t.templateElement({ raw: `@replexica/translations/`, cooked: `@replexica/translations/` }, false),
                    t.templateElement({ raw: '.json', cooked: '.json' }, true),
                  ], [t.identifier('locale')])],
                ),
              ),
            ),
          )
        );
      }

      chunk.path.replaceWith(
        t.jsxElement(
          injectedElement,
          null,
          [],
          true,
        )  
      );

      result[chunk.id] = chunk.text;
    }

    return result;
  }

  public extractHints(): ReplexicaScopeHint[] {    
    const result = this._extractBaseHints(this.path);
    return result;
  }
}
