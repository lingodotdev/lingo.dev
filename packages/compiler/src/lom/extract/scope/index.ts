import { composeParsers } from "../_utils";
import { fromJsxElement } from "./parsers";

export default composeParsers(
  'scope',
  fromJsxElement,
);
