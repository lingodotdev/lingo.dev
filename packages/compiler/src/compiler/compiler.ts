import * as t from "@babel/types";
import { traverse } from '@babel/core';
import { generateCodeFromBabelAst, parseCodeIntoBabelAst } from "../utils/babel";
import { generateFileId } from "../utils/id";
import { GeneratorResult } from "@babel/generator";
import { hasDirective } from "../utils/ast";
import { ReplexicaFileData, ReplexicaCompilerData } from "./types";
import { ReplexicaScopeExtractor } from "./scope";

export class ReplexicaCompiler {
  public static fromCode(code: string, relativeFilePath: string, rsc: boolean) {
    const ast = parseCodeIntoBabelAst(code);
    const isServer = !rsc ? false : !hasDirective(ast, 'use client');

    return new ReplexicaCompiler(relativeFilePath, code, ast, isServer);
  }

  private constructor(
    public readonly relativeFilePath: string,
    private readonly code: string,
    public readonly ast: t.File,
    private readonly isServer: boolean,
  ) {
    this.fileId = generateFileId(relativeFilePath, 0);
  }

  private readonly fileId: string;

  private _extractors: Set<ReplexicaScopeExtractor> = new Set();
  private _data: ReplexicaFileData = {};

  public withScope(extractor: ReplexicaScopeExtractor): this {
    this._extractors.add(extractor);
    return this;
  }

  public injectIntl(): this {
    const _compiler = this;
    traverse(this.ast, {
      enter(path) {
        try {
          for (const extractor of _compiler._extractors) {
            const scopes = extractor.fromNode(path);
            for (const scope of scopes) {
              const hints = scope.extractHints();
              const data = scope.injectIntl(_compiler.fileId, _compiler.isServer);
              _compiler._data[scope.id] = { hints, data };
            }
          }
        } catch (thrownObject) {
          if (thrownObject === 'skip') {
            path.skip();
          } else {
            throw thrownObject;
          }
        }
      }
    });
    return this;
  }

  public get data(): ReplexicaCompilerData {
    return {
      [this.fileId]: {
        context: { isClient: !this.isServer },
        data: this._data,
      },
    };
  }

  public generate(): GeneratorResult {
    return generateCodeFromBabelAst(this.code, this.ast);
  }
}
