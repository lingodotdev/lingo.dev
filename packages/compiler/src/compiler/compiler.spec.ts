import { describe, it, expect, vi, beforeEach } from 'vitest';

import { ReplexicaCompiler } from './compiler';
import { ReplexicaAttributeScope, ReplexicaContentScope, ReplexicaSkipScope } from './scope';

const mocks = vi.hoisted(() => {
  return {
    generateScopeId: vi.fn(),
    generateChunkId: vi.fn(),
    generateFileId: vi.fn(),
  };
});

vi.mock('../utils/id', () => {
  return {
    generateScopeId: mocks.generateScopeId,
    generateChunkId: mocks.generateChunkId,
    generateFileId: mocks.generateFileId,
  };
});

describe('Compiles content', () => {
  beforeEach(() => {
    let fileIdCounter = 0;
    let scopeIdCounter = 0;
    let chunkIdCounter = 0;

    mocks.generateFileId.mockImplementation(() => `fileId_${fileIdCounter++}`);
    mocks.generateScopeId.mockImplementation(() => `scopeId_${scopeIdCounter++}`);
    mocks.generateChunkId.mockImplementation(() => `chunkId_${chunkIdCounter++}`);
  });

  it('one basic span', () => {
    const inputCode = '<span>Hello, world!</span>';
    const compiler = createCompiler(inputCode);

    const outputCode = compiler
      .injectIntl()
      .generate()
      .code;
    const outputData = compiler.data;

    expect(outputCode).toMatchSnapshot();
    expect(outputData).toEqual({
      fileId_0: {
        context: { isClient: false },
        data: {
          scopeId_0: {
            hints: [{
              type: 'element',
              name: 'span',
              hint: null,
            }],
            data: {
              chunkId_0: 'Hello, world!',
            },
          },
        },
      },
    });
  });

  it('multiple basic spans', () => {    
    const inputCode = `
<div>
  <span>Hello, world!</span>
  <span>Goodbye, world!</span>
</div>
    `.trim();

    const compiler = createCompiler(inputCode);

    const outputCode = compiler
      .injectIntl()
      .generate()
      .code;
    const outputData = compiler.data;

    expect(outputCode).toMatchSnapshot();
    expect(outputData).toEqual({
      fileId_0: {
        context: { isClient: false },
        data: {
          scopeId_0: {
            hints: [{
              type: 'element',
              name: 'span',
              hint: null,
            }],
            data: {
              chunkId_0: 'Hello, world!',
            },
          },
          scopeId_1: {
            hints: [{
              type: 'element',
              name: 'span',
              hint: null,
            }],
            data: {
              chunkId_1: 'Goodbye, world!',
            },
          },
        },
      },
    });
  });

  it('nested basic spans', () => {
    const inputCode = `
<div>
  <span>Hello, world!</span>
  <div>
    <span>Goodbye, world!</span>
  </div>
</div>
    `.trim();

    const compiler = createCompiler(inputCode);

    const outputCode = compiler
      .injectIntl()
      .generate()
      .code;
    const outputData = compiler.data;

    expect(outputCode).toMatchSnapshot();
    expect(outputData).toEqual({
      fileId_0: {
        context: { isClient: false },
        data: {
          scopeId_0: {
            hints: [{
              type: 'element',
              name: 'span',
              hint: null,
            }],
            data: {
              chunkId_0: 'Hello, world!',
            },
          },
          scopeId_1: {
            hints: [{
              type: 'element',
              name: 'span',
              hint: null,
            }],
            data: {
              chunkId_1: 'Goodbye, world!',
            },
          },
        },
      },
    });
  });

  it('nested elements mixed with text', () => {
    const inputCode = `
<div>
  <span>Hello, world!</span>
  <div>
    Goodbye, world!
    <br />
    You can say <b>hello</b> again by clicking <a href="#">here</a>.
  </div>
</div>
    `.trim();
    const compiler = createCompiler(inputCode);

    const outputCode = compiler
      .injectIntl()
      .generate()
      .code;
    const actualData = compiler.data;

    expect(outputCode).toMatchSnapshot();
    expect(actualData).toEqual({
      fileId_0: {
        context: { isClient: false },
        data: {
          scopeId_0: {
            hints: [{
              type: 'element',
              name: 'span',
              hint: null,
            }],
            data: {
              chunkId_0: 'Hello, world!',
            },
          },
          scopeId_1: {
            hints: [{
              type: 'element',
              name: 'div',
              hint: null,
            }],
            data: {
              chunkId_1: ' Goodbye, world! ',
              chunkId_2: ' You can say ',
              chunkId_3: 'hello',
              chunkId_4: ' again by clicking ',
              chunkId_5: 'here',
              chunkId_6: '. ',
            },
          },
        },
      },
    });
  });

  it('jsx assigned to jsx attribute', () => {
    const inputCode = `
<div>
  <div
    data-test={<span>Hello, world!</span>}
  >
    Goodbye, world!
  </div>
</div>
    `.trim();
    const compiler = createCompiler(inputCode);

    const outputCode = compiler
      .injectIntl()
      .generate()
      .code;
    const actualData = compiler.data;

    expect(outputCode).toMatchSnapshot();
    expect(actualData).toEqual({
      fileId_0: {
        context: { isClient: false },
        data: {
          scopeId_0: {
            hints: [{
              type: 'element',
              name: 'div',
              hint: null,
            }],
            data: {
              chunkId_0: ' Goodbye, world! ',
            },
          },
          scopeId_1: {
            hints: [{
              type: 'element',
              name: 'span',
              hint: null,
            }],
            data: {
              chunkId_1: 'Hello, world!',
            },
          },
        },
      },
    });
  });
});

describe('Skips nodes marked with `data-replexica-skip`', () => {
  beforeEach(() => {
    let fileIdCounter = 0;
    let scopeIdCounter = 0;
    let chunkIdCounter = 0;

    mocks.generateFileId.mockImplementation(() => `fileId_${fileIdCounter++}`);
    mocks.generateScopeId.mockImplementation(() => `scopeId_${scopeIdCounter++}`);
    mocks.generateChunkId.mockImplementation(() => `chunkId_${chunkIdCounter++}`);
  });

  it('skips a single node', () => {
    const inputCode = `
<div>
  <span>Hello, world!</span>
  <div data-replexica-skip>
    Goodbye, world!
  </div>
</div>
    `.trim();
    const compiler = createCompiler(inputCode);

    const outputCode = compiler
      .injectIntl()
      .generate()
      .code;
    const actualData = compiler.data;

    expect(outputCode).toMatchSnapshot();
    expect(actualData).toEqual({
      fileId_0: {
        context: { isClient: false },
        data: {
          scopeId_0: {
            hints: [{
              type: 'element',
              name: 'span',
              hint: null,
            }],
            data: {
              chunkId_0: 'Hello, world!',
            },
          },
        },
      },
    });
  });

  it('skips entire subtree', () => {
    const inputCode = `
<div>
  <span>Hello, world!</span>
  <div data-replexica-skip>
    Goodbye, world!
    <br />
    You can say <b>hello</b> again by clicking <a href="#">here</a>.
  </div>
</div>
    `.trim();
    const compiler = createCompiler(inputCode);

    const outputCode = compiler
      .injectIntl()
      .generate()
      .code;
    const actualData = compiler.data;

    expect(outputCode).toMatchSnapshot();
    expect(actualData).toEqual({
      fileId_0: {
        context: { isClient: false },
        data: {
          scopeId_0: {
            hints: [{
              type: 'element',
              name: 'span',
              hint: null,
            }],
            data: {
              chunkId_0: 'Hello, world!',
            },
          },
        },
      },
    });
  });
});

describe('Compiles attribute values', () => {
  beforeEach(() => {
    let fileIdCounter = 0;
    let scopeIdCounter = 0;
    let chunkIdCounter = 0;

    mocks.generateFileId.mockImplementation(() => `fileId_${fileIdCounter++}`);
    mocks.generateScopeId.mockImplementation(() => `scopeId_${scopeIdCounter++}`);
    mocks.generateChunkId.mockImplementation(() => `chunkId_${chunkIdCounter++}`);
  });

  it('compiles attribute values', () => {
    const inputCode = `
<div>
  <span title="Hello!">World</span>
</div>
    `;
    const compiler = createCompiler(inputCode);

    const outputCode = compiler
      .injectIntl()
      .generate()
      .code;
    const actualData = compiler.data;

    expect(outputCode).toMatchSnapshot();
    expect(actualData).toEqual({
      fileId_0: {
        context: { isClient: false },
        data: {
          scopeId_0: {
            hints: [{
              type: 'element',
              name: 'span',
              hint: null,
            }],
            data: {
              chunkId_0: 'World',
            },
          },
          scopeId_1: {
            hints: [{
              type: 'element',
              name: 'span',
              hint: null,
            }, {
              type: 'attribute',
              name: 'title',
            }],
            data: {
              chunkId_1: 'Hello!',
            },
          },
        },
      },
    });
  });
});

describe('Adds hints added explicitly with `data-replexica-hint`', () => {
  beforeEach(() => {
    let fileIdCounter = 0;
    let scopeIdCounter = 0;
    let chunkIdCounter = 0;

    mocks.generateFileId.mockImplementation(() => `fileId_${fileIdCounter++}`);
    mocks.generateScopeId.mockImplementation(() => `scopeId_${scopeIdCounter++}`);
    mocks.generateChunkId.mockImplementation(() => `chunkId_${chunkIdCounter++}`);
  });

  it('simple nested span', () => {
    const inputCode = `
<div>
  <span data-replexica-hint="hello world message">Hello, world!</span>
</div>
    `;
    const compiler = createCompiler(inputCode);

    compiler.injectIntl();
    const actualData = compiler.data;

    expect(actualData).toEqual({
      fileId_0: {
        context: { isClient: false },
        data: {
          scopeId_0: {
            hints: [{
              type: 'element',
              name: 'span',
              hint: 'hello world message',
            }],
            data: {
              chunkId_0: 'Hello, world!',
            },
          },
        },
      },
    });
  });

  it('nested span with an attribute', () => {
    const inputCode = `
<div>
  <span title="Hello!" data-replexica-hint="hello world message">World</span>
</div>
    `;
    const compiler = createCompiler(inputCode);

    compiler.injectIntl();
    const actualData = compiler.data;

    expect(actualData).toEqual({
      fileId_0: {
        context: { isClient: false },
        data: {
          scopeId_0: {
            hints: [{
              type: 'element',
              name: 'span',
              hint: 'hello world message',
            }],
            data: {
              chunkId_0: 'World',
            },
          },
          scopeId_1: {
            hints: [{
              type: 'element',
              name: 'span',
              hint: 'hello world message',
            }, {
              type: 'attribute',
              name: 'title',
            }],
            data: {
              chunkId_1: 'Hello!',
            },
          },
        },
      },
    });

  });
});

// helpers

function createCompiler(inputCode: string) {
  return ReplexicaCompiler
    .fromCode(inputCode, '/path/to/file.tsx', true)
    .withScope(ReplexicaSkipScope)
    .withScope(ReplexicaContentScope)
    .withScope(ReplexicaAttributeScope)
  ;
}