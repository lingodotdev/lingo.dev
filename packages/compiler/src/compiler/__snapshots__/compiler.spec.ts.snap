// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Compiles attribute values > compiles attribute values 1`] = `
"import { I18nProxy as _I18nProxy } from "@replexica/react/server";
import { loadLocaleData as _loadLocaleData } from "@replexica/react/next";
import { loadLocaleFromCookie as _loadLocaleFromCookie } from "@replexica/react/next";
import { I18nChunk as _I18nChunk } from "@replexica/react/server";
<div>
  <_I18nProxy title="Hello!" $$Component="span" $$Attributes={{
    "title": {
      "fileId": "fileId_0",
      "scopeId": "scopeId_1",
      "chunkId": "chunkId_1"
    }
  }} loadLocale={_loadLocaleFromCookie} loadLocaleData={locale => import(\`@replexica/translations/\${locale}.json\`)}><_I18nChunk fileId="fileId_0" scopeId="scopeId_0" chunkId="chunkId_0" loadLocale={_loadLocaleFromCookie} loadLocaleData={locale => import(\`@replexica/translations/\${locale}.json\`)} /></_I18nProxy>
</div>;"
`;

exports[`Compiles content > jsx assigned to jsx attribute 1`] = `
"import { loadLocaleData as _loadLocaleData } from "@replexica/react/next";
import { loadLocaleFromCookie as _loadLocaleFromCookie } from "@replexica/react/next";
import { I18nChunk as _I18nChunk } from "@replexica/react/server";
<div>
  <div data-test={<span><_I18nChunk fileId="fileId_0" scopeId="scopeId_1" chunkId="chunkId_1" loadLocale={_loadLocaleFromCookie} loadLocaleData={locale => import(\`@replexica/translations/\${locale}.json\`)} /></span>}><_I18nChunk fileId="fileId_0" scopeId="scopeId_0" chunkId="chunkId_0" loadLocale={_loadLocaleFromCookie} loadLocaleData={locale => import(\`@replexica/translations/\${locale}.json\`)} /></div>
</div>;"
`;

exports[`Compiles content > multiple basic spans 1`] = `
"import { loadLocaleData as _loadLocaleData } from "@replexica/react/next";
import { loadLocaleFromCookie as _loadLocaleFromCookie } from "@replexica/react/next";
import { I18nChunk as _I18nChunk } from "@replexica/react/server";
<div>
  <span><_I18nChunk fileId="fileId_0" scopeId="scopeId_0" chunkId="chunkId_0" loadLocale={_loadLocaleFromCookie} loadLocaleData={locale => import(\`@replexica/translations/\${locale}.json\`)} /></span>
  <span><_I18nChunk fileId="fileId_0" scopeId="scopeId_1" chunkId="chunkId_1" loadLocale={_loadLocaleFromCookie} loadLocaleData={locale => import(\`@replexica/translations/\${locale}.json\`)} /></span>
</div>;"
`;

exports[`Compiles content > nested basic spans 1`] = `
"import { loadLocaleData as _loadLocaleData } from "@replexica/react/next";
import { loadLocaleFromCookie as _loadLocaleFromCookie } from "@replexica/react/next";
import { I18nChunk as _I18nChunk } from "@replexica/react/server";
<div>
  <span><_I18nChunk fileId="fileId_0" scopeId="scopeId_0" chunkId="chunkId_0" loadLocale={_loadLocaleFromCookie} loadLocaleData={locale => import(\`@replexica/translations/\${locale}.json\`)} /></span>
  <div>
    <span><_I18nChunk fileId="fileId_0" scopeId="scopeId_1" chunkId="chunkId_1" loadLocale={_loadLocaleFromCookie} loadLocaleData={locale => import(\`@replexica/translations/\${locale}.json\`)} /></span>
  </div>
</div>;"
`;

exports[`Compiles content > nested elements mixed with text 1`] = `
"import { loadLocaleData as _loadLocaleData } from "@replexica/react/next";
import { loadLocaleFromCookie as _loadLocaleFromCookie } from "@replexica/react/next";
import { I18nChunk as _I18nChunk } from "@replexica/react/server";
<div>
  <span><_I18nChunk fileId="fileId_0" scopeId="scopeId_0" chunkId="chunkId_0" loadLocale={_loadLocaleFromCookie} loadLocaleData={locale => import(\`@replexica/translations/\${locale}.json\`)} /></span>
  <div><_I18nChunk fileId="fileId_0" scopeId="scopeId_1" chunkId="chunkId_1" loadLocale={_loadLocaleFromCookie} loadLocaleData={locale => import(\`@replexica/translations/\${locale}.json\`)} /><br /><_I18nChunk fileId="fileId_0" scopeId="scopeId_1" chunkId="chunkId_2" loadLocale={_loadLocaleFromCookie} loadLocaleData={locale => import(\`@replexica/translations/\${locale}.json\`)} /><b><_I18nChunk fileId="fileId_0" scopeId="scopeId_1" chunkId="chunkId_3" loadLocale={_loadLocaleFromCookie} loadLocaleData={locale => import(\`@replexica/translations/\${locale}.json\`)} /></b><_I18nChunk fileId="fileId_0" scopeId="scopeId_1" chunkId="chunkId_4" loadLocale={_loadLocaleFromCookie} loadLocaleData={locale => import(\`@replexica/translations/\${locale}.json\`)} /><a href="#"><_I18nChunk fileId="fileId_0" scopeId="scopeId_1" chunkId="chunkId_5" loadLocale={_loadLocaleFromCookie} loadLocaleData={locale => import(\`@replexica/translations/\${locale}.json\`)} /></a><_I18nChunk fileId="fileId_0" scopeId="scopeId_1" chunkId="chunkId_6" loadLocale={_loadLocaleFromCookie} loadLocaleData={locale => import(\`@replexica/translations/\${locale}.json\`)} /></div>
</div>;"
`;

exports[`Compiles content > one basic span 1`] = `
"import { loadLocaleData as _loadLocaleData } from "@replexica/react/next";
import { loadLocaleFromCookie as _loadLocaleFromCookie } from "@replexica/react/next";
import { I18nChunk as _I18nChunk } from "@replexica/react/server";
<span><_I18nChunk fileId="fileId_0" scopeId="scopeId_0" chunkId="chunkId_0" loadLocale={_loadLocaleFromCookie} loadLocaleData={locale => import(\`@replexica/translations/\${locale}.json\`)} /></span>;"
`;

exports[`Skips nodes marked with \`data-replexica-skip\` > skips a single node 1`] = `
"import { loadLocaleData as _loadLocaleData } from "@replexica/react/next";
import { loadLocaleFromCookie as _loadLocaleFromCookie } from "@replexica/react/next";
import { I18nChunk as _I18nChunk } from "@replexica/react/server";
<div>
  <span><_I18nChunk fileId="fileId_0" scopeId="scopeId_0" chunkId="chunkId_0" loadLocale={_loadLocaleFromCookie} loadLocaleData={locale => import(\`@replexica/translations/\${locale}.json\`)} /></span>
  <div data-replexica-skip>
    Goodbye, world!
  </div>
</div>;"
`;

exports[`Skips nodes marked with \`data-replexica-skip\` > skips entire subtree 1`] = `
"import { loadLocaleData as _loadLocaleData } from "@replexica/react/next";
import { loadLocaleFromCookie as _loadLocaleFromCookie } from "@replexica/react/next";
import { I18nChunk as _I18nChunk } from "@replexica/react/server";
<div>
  <span><_I18nChunk fileId="fileId_0" scopeId="scopeId_0" chunkId="chunkId_0" loadLocale={_loadLocaleFromCookie} loadLocaleData={locale => import(\`@replexica/translations/\${locale}.json\`)} /></span>
  <div data-replexica-skip>
    Goodbye, world!
    <br />
    You can say <b>hello</b> again by clicking <a href="#">here</a>.
  </div>
</div>;"
`;
